src/NumericalAlgorithms/Interpolation/LinearInterpolationND.hpp

```diff
@@ -0,0 +1,374 @@
+ public:
+  using DataPointer = gsl::span<double const>;
+  //! X values of the table. Only used if allocated
+  std::array<DataPointer, Dimension> x;
+  //! Y values of the table. Only used if allocated
+  DataPointer y;
+
+  std::array<size_t, Dimension> numberOfPoints;
+  std::array<double, Dimension> inverseSpacing, spacing;
+
+  std::array<bool, Dimension> extrapolate_low;
+  std::array<bool, Dimension> extrapolate_high;
+
+ private:
+  //! Determine relative index bracketing for non-uniform spacing
+  inline size_t findIndexGeneral(size_t const whichDimension,
+                                 double const& xin) const {
+    int lowerIndexBound = 0;
+    int upperIndexBound = numberOfPoints[whichDimension] - 1;
+
+    ASSERT((xin > x[whichDimension][lowerIndexBound]) or
+               extrapolate_low[whichDimension],
+           "Interpolation exceeds lower table bounds");
+    ASSERT((xin < x[whichDimension][upperIndexBound]) or
+               extrapolate_high[whichDimension],
+           "Interpolation exceeds upper table bounds");
+
+    while (upperIndexBound - lowerIndexBound > 1) {
+      int currentIndex =
+          lowerIndexBound + (upperIndexBound - lowerIndexBound) / 2;
+      if (xin < x[whichDimension][currentIndex])
+        upperIndexBound = currentIndex;
+      else
+        lowerIndexBound = currentIndex;
+    }
+    return lowerIndexBound;
+  };
+
+  //! Determine relative index bracketing for non-uniform spacing
+  inline size_t findIndexUniform(size_t const whichDimension,
+                                 double const& xin) const {
+    // Compute coordinate relative to lowest table bound
+    const auto relativeCoordinate = (xin - x[whichDimension][0]);
+
+    if (relativeCoordinate == 0) {
+      return 0;
+    }
+
+    // We are exceeding the table bounds:
+    // Use linear extrapolation based of the lowest
+    // two points in the table
+    if (relativeCoordinate < 0) {
+      ASSERT(extrapolate_low[whichDimension],
+             "Interpolation exceeds lower table bounds.");
+    }
+
+    auto currentIndex =
+        static_cast<int>(relativeCoordinate * inverseSpacing[whichDimension]);
+
+    // We are exceeding the table bounds:
+    // Use linear extrapolation based of the highest
+    // two points in the table
+
+    if (currentIndex > numberOfPoints[whichDimension] - 2) {
+      ASSERT(extrapolate_high[whichDimension],
+             "Interpolation exceeds upper table bounds.");
+
+      return numberOfPoints[whichDimension] - 2;
+    }
+
+    return currentIndex;
+  };
+
+  inline size_t findIndex(size_t const whichDimension,
+                          double const& xin) const {
+    if constexpr (UniformSpacing) {
+      return findIndexUniform(whichDimension, xin);
+    } else {
+      return findIndexGeneral(whichDimension, xin);
+    }
+  };
+
+  //! Recursive interpolaton requires computing compressed indices.
+  //  We assume the table layout to be in Fortran ordering.
```

Comment:
`Fortran ordering (last index varies fastest)`

This should also be on the main documentation.